src/main/kotlin/org/rust/ide/annotator/fixes/ChangeFunctionSignatureFix.kt

```diff
@@ -217,25 +331,28 @@ private class PeekableIterator<T>(private val iterator: Iterator<T>) {
     }
 }
 
-private sealed class SignatureMember {
-    data class Parameter(val parameter: RsValueParameter) : SignatureMember()
-    data class Argument(val expr: RsExpr) : SignatureMember()
+private data class Signature(val actions: List<SignatureAction>) {
+    /**
+     * Calculates how many parameters would a function had if these actions were applied to its signature.
+     */
+    fun getActualParameterCount(): Int = actions.sumOf {
+        val count = when (it) {
+            is SignatureAction.RemoveParameter -> -1
+            else -> 1
+        }
+        count
+    }
 }
 
-private fun suffix(number: Int): String {
-    if ((number % 100) in 11..13) {
-        return "th"
-    }
-    return when (number % 10) {
-        1 -> "st"
-        2 -> "nd"
-        3 -> "rd"
-        else -> "th"
-    }
+private sealed class SignatureAction {
```

Comment:
Do I understand correctly that we take desired signature (based on arguments types), create diff using this class by comparing desired signature with current signature, and then apply diff to function? Wouldn't it be easier if we take desired signature and directly modify function signature (without creating diff using `Keep`/`Insert`/`Remove` actions)?

Reply:
It's a list of actions that can be taken to reach some target signature. Even though a target signature might be the same, you can reach it with different actions. This was also a "problem" in the `Change signature` refactoring, which is used by this fix. In the refactoring I need to be able to identify e.g. if the user moved a parameter or deleted it and then created a new one, and here the situation is the same.

A very simple example:
```
source signature: (u32, u32)
target_signature: (u32, u32, u32)

actions_1: (keep, insert, keep)
actions_2: (keep, keep, insert)
```
Just storing the target signature would lose information about how to reach it. The actions affect the wording of the fix and also the semantics (as we have discussed a few comments above). I need to be able to trace the actions that have to be performed to change the signature, otherwise the `Change signature` refactoring will not work properly.

An action at index `i` says how to build the `i-th` member of the target signature:
1) KeepParameter(source parameter) -> use the parameter from a specified index the source signature
2) InsertArgument(source argument) -> create a new parameter from the specified argument
3) RemoveParameter -> remove the parameter
4) ChangeParameterType(source argument) -> change the parameter type, based on the specified argument

The `RemoveParameter` differs from the others, but there's an invariant that it can only be created at the end of the action list. This makes the implementation simpler, otherwise removing from the middle would complicate stuff a lot, but we do not need that.

Reply:
Ok, it is reasonable, thanks for the explanation!